utils.py

```diff
@@ -0,0 +1,143 @@
+import time
+import streamlit as st
+import pandas as pd
+import os
+import json
+import pycld2 as cld2
+import regex
+
+
+def time_it(label, fn):
+    """
+    Wrap a function with a timer
+
+    label: string label or method that creates a label from the result of calling fn
+    fn: the function to wrap
+    """
+
+    def wrapper(*args):
+        start_time = time.perf_counter()
+        result = fn(*args)
+        end_time = time.perf_counter()
+        label_str = str(label(result)) if callable(label) else str(label)
+
+        print("[Timer]: {} took {:.3f}s".format(label_str, end_time - start_time))
+        return result
+
+    return wrapper
+
+
+def one_hot_encode_authors(df):
+    author_cols = [col for col in df if col.startswith("Author_")]
+    df = pd.get_dummies(df, columns=author_cols, sparse=True, prefix="Author")
+    return df
+
+
+def st_dataset_selector():
+    # Enables streamlit to read data from mounted disk in GCP
+    dataset_location = "./"
+    if "STREAMLIT_DATA_LOCATION" in os.environ:
+        dataset_location = os.environ["STREAMLIT_DATA_LOCATION"]
+
+    datasets = {
+        dataset_location + "sample_data.jsonl": "Small sample (50 rows)",
+        dataset_location + "250k.docs.jsonl": "Large sample (250k rows)",
+        dataset_location + "mag5.docs.jsonl": "Full dataset (5m rows, slooow)",
+    }
+
+    return st.selectbox(
+        "Selected Dataset",
+        list(datasets.keys()),
+        format_func=lambda x: datasets[x],
+    )
+
+
+RE_BAD_CHARS = regex.compile(r"\p{Cc}|\p{Cs}")
+
+
+# Due to https://github.com/mikemccand/chromium-compact-language-detector/issues/22
+# cld2 can't handle unprintable characters
+def strip_unprintable(s):
+    """
+    Strip non-printable characters
+    """
+    return "".join(c for c in s if c.isprintable())
+
+
+@st.cache(suppress_st_warning=True)
+def load_dataset(dataset_filename, limit):
```

Comment:
Do we have a custom method just for the progress bar? What's the difference over pd.read_json?

Reply:
This method reads each json line in the file and creates a dataframe. It is different from `pd.read_json`, since it handles nested json fields/array fields, deletes some irrelevant fields & strips irrelevant characters. 

Using `pd.read_json` on the whole json file would be slower and more memory-intensive since we would need to read the whole file into memory and then handle each column after.